### PR TITLE
[pack] Include Extension and Dynamic Concurrency Settings in Sync Triggers Payload

### PR DESCRIPTION
**Pull request by glennamanns**
Resolves #https://github.com/Azure/azure-functions-host/issues/7394

@TsuyoshiUshio is driving this work
I'd like to list several work items for this PR has been merged. Originally this PR is not aimed for being merged, so that I need refactoring to make it production.  

CC @pragnagopa  @chiangvincent 

- [ ] Test with ScaleController for the current implement with e2e 
- [ ] Discuss how we implement this change with ArmChacheEnabled with @mathewc  I don't know what is it. Any pointer?
- [ ] Fix the integration testing issue (Currently a lot of testing fails)  with refactoring Logs. 
- [ ] Add test  cases for the Integration testing
- [ ] Add Validation of the Payload size
- [ ] Find out where the SyncTrigger payload is used? As far as I know, 1. [k4apps](https://github.com/Azure-App-Service/KuduLite/blob/6d5185b28bcf9efb88f26a098b4fc806ff17d534/Kudu.Core/Functions/KedaFunctionTriggerProvider.cs#L115)  2. FrontEnd (I can see the code of SyncTrigger) / ScaleController **Do we have any other places to check the compatibility?** @mathewc 
- [ ] Test again with ScaleController/FrontEnd and k4apps  (or other places)

### CODE REVIEW

**Comment by pragnagopa** [comment]
@glennamanns - do we need this PR for Dynamic Concurrency + Scale Controller updates?

**Comment by pragnagopa** [comment]
cc @TsuyoshiUshio @chiangvincent

**Comment by TsuyoshiUshio** [comment]
https://github.com/Azure/azure-functions-host/issues/7394 this PR solves this issue.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs [hunk 1]
```diff
@@ -306,13 +314,26 @@ public async Task<SyncTriggersPayload> GetSyncTriggersPayload()
...
+                result.Add("extensions", extensionsPayload);
+            }
+
+            JObject concurrencyPayload = GetHostJsonSection(jsonPayload, "concurrency", _logger);
+            if (concurrencyPayload != null)
+            {
+                result.Add("concurrency", concurrencyPayload);
+            }
+
+            JObject kedaPayload = GetHostJsonSection(jsonPayload, "kedaScalingAlgorithm", _logger);
```
Comment (TsuyoshiUshio) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: This section is not necessary, the keda style algorithm will be defined by AppSettings.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs [hunk 2]
```diff
@@ -306,13 +314,26 @@ public async Task<SyncTriggersPayload> GetSyncTriggersPayload()
...
+            if (kedaPayload != null)
+            {
+                result.Add("kedaScalingAlgorithm", kedaPayload);
```
Comment (TsuyoshiUshio) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Same as avobe.

**Review by TsuyoshiUshio — COMMENTED** [review]
We need to do three steps. The code looks good except for my comment, however, we need to do the testing for these scenario to accept it. 
1. Test against Scale Controller on a private stamp
2. Test against Arc on a AKS cluster if the transformation is success. 
Once the test has done, we can remove the Logging from [Glenna] part.

**Comment by TsuyoshiUshio** [comment]
https://github.com/Azure/azure-functions-host/issues/7299

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs [hunk 3]
```diff
@@ -285,12 +298,7 @@ public async Task<SyncTriggersPayload> GetSyncTriggersPayload()
 
             if (!ArmCacheEnabled)
```
Comment (TsuyoshiUshio) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:326]: @pragnagopa  Do you know what is the ArmCacheEnabled?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs (hunk 1 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Removed

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs (hunk 2 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Removed

**Comment by TsuyoshiUshio** [comment]
I'd like to list several work items for this PR has been merged. Originally this PR is not aimed for being merged, so that I need refactoring to make it production.  CC @pragnagopa  @chiangvincent 

- [ ] Test with ScaleController for the current implement with e2e 
- [ ] Discuss how we implement this change with ArmChacheEnabled with @mathewc  I don't know what is it.
- [ ] Fix the integration testing issue (Currently a lot of testing fails) 
- [ ] Add test  cases for the Integration testing
- [ ] Find out where the SyncTrigger payload is used? As far as I know, 1. [k4apps](https://github.com/Azure-App-Service/KuduLite/blob/6d5185b28bcf9efb88f26a098b4fc806ff17d534/Kudu.Core/Functions/KedaFunctionTriggerProvider.cs#L115)  2. FrontEnd (I can see the code of SyncTrigger) / ScaleController Do we have any other places to check the compatibility?
- [ ] Test again with ScaleController/FrontEnd and k4apps  (or other places)

**Comment by TsuyoshiUshio** [comment]
Currently, public async Task TrySyncTriggers_MaxSyncTriggersPayloadSize_Succeeds() fails since we don't implement ArmCacheEnabled before we can talk the spec. https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=41208&view=ms.vss-test-web.build-test-results-tab&runId=1160974&resultId=100212&paneView=debug

**Comment by pragnagopa** [comment]
@TsuyoshiUshio - please clean up the PR - fix logging to help start with reviews

**Comment by mathewc** [comment]
Recommend closing this test PR and opening a new PR with production code. In general that PR will:

1. Add the "extensions" and "concurrency" sections to the sync triggers payload in both full and minimal formats. I.e. the minimal format returned when ArmCacheEnabled is true or when the full payload is oversize will now be an object format always including these two properties in addition to "triggers"
2. For minimal format, we'll no longer return an array format - we'll always return object. That ensures we can always include these properties. ScaleController and other consumers already handle both array and object formats of course.
3. Remove the Kubernetes specific changes added in https://github.com/Azure/azure-functions-host/pull/7386/files
4. Once ScaleController is updated to consume these changes, we can finally address https://github.com/Azure/azure-functions-host/issues/5390 by removing all the Durable specific code from the host. SC will have what it needs at that point.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs (hunk 3 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:326]: The extended sync triggers object format payload was added as part of some Functions ARM optimization work, We started including "functions" and "secrets" details in the payload, which enables us to serve common ARM requests from cached data improving API performance. When that change was introduced, we included config flags for safety to enable users to revert if they saw issues.

**Comment by pragnagopa** [comment]
Thanks @mathewc  for the feedback. @TsuyoshiUshio  - Lets go ahead and start a new PR as the changes will address more generic requirements.

**Comment by liliankasem** [comment]
> Thanks @mathewc for the feedback. @TsuyoshiUshio - Lets go ahead and start a new PR as the changes will address more generic requirements.

Is it safe to close this PR?

**Comment by liliankasem** [comment]
Looks like #7863 is addressing this, closing PR. If we need to track work that needs to be done, let's open an issue instead. If necessary we can always reopen.

**Pull request closed without merging**